src/tool_provider/artifactory.rs

```diff
@@ -18,19 +26,120 @@ impl ArtifactoryProvider {
         Self { paths }
     }
 }
-#[allow(unused)]
+
 impl ToolProviderImpl for ArtifactoryProvider {
-    fn get_releases(&self, repo: &str) -> ForemanResult<Vec<Release>> {
-        Err(ForemanError::Other {
-            message: "Artifactory is not yet supported. Please use Github or Gitlab as your source"
-                .to_owned(),
-        })
+    fn get_releases(&self, repo: &str, host: &Url) -> ForemanResult<Vec<Release>> {
+        let client = Client::new();
+
+        let url = format!("{}artifactory/api/storage/{}", host, repo);
+        let params = vec![("list", ""), ("deep", "1")];
+        let mut builder = client
+            .get(&url)
+            .header(USER_AGENT, "Roblox/foreman")
+            .query(&params);
+
+        let tokens = artiaa_auth::Tokens::load(&self.paths.artiaa_path()?)
+            .map_err(|error| ForemanError::ArtiAAError { error })?;
+
+        if let Some(credentials) = tokens.get_credentials(host) {
+            builder = builder.header(ARTIFACTORY_API_KEY_HEADER, credentials.token.to_string());
+        }
+        log::debug!("Downloading artifactory releases for {}", repo);
+        let response_body = builder
+            .send()
+            .map_err(ForemanError::request_failed)?
+            .text()
+            .map_err(ForemanError::request_failed)?;
+
+        let response: ArtifactoryResponse =
+            serde_json::from_str(&response_body).map_err(|err| {
+                ForemanError::unexpected_response_body(err.to_string(), response_body, url)
+            })?;
+
+        let mut releases: Vec<ArtifactoryRelease> = Vec::new();
+        let mut release_map: HashMap<&str, Vec<ArtifactoryAsset>> = HashMap::new();
+        for file in &response.files {
+            let mut uri = file.uri.split("/");
+            // file.uri will look something like /<version>/<artifact-name>, so uri will be ["", <version>, <artifact-name]
+            uri.next();
+            let version = uri.next().unwrap();
+            let asset_name = uri.next().unwrap();
```

Comment:
Should we do some validation here? For names, I don't think they have to be especially precise, but we could certainly validate that `version` is a valid semver version, right?

Reply:
I guess we probably do this elsewhere, now that I think about it. But I do wonder if we could detect if the target repo isn't structured the way we expect.

Reply:
Good point. I think it's worth at least ensuring that the repo path is structured the way we expect. If the path isn't what we expect, we should probably skip the file instead of panicing though.

Reply:
Foreman already does this version validation here https://github.com/Roblox/foreman/blob/main/src/tool_cache.rs#L115